Add code activity graph to project widget

sebastinez opened the pull request:
Already working but there are some optimizations left to do.

- General refactoring to make the code simpler.
- Improve the display of the graphs (we have some graphs of peaks where 1 or 2 commits seem the same as 100+ commits in a week)

sebastinez commented:
Ok it's ready, I think it's stable enough now.
The only issue is that e.g. one commit in a new project gets a massive spike while a project with many commits per week seem less spiky..

I think there is a solution there to distribute the activity graphs evenly between all projects, but I think thats for another PR.

cloudhead commented on src/commit.ts:
```diff
@@ -108,3 +111,57 @@ export function groupCommits(commits: CommitHeader[]): CommitGroup[] {
+    }
+
+    return 0;
+  });
```
Note that this is sorting in reverse order, ie. from latest commit to earliest. Is this intended?

sebastinez replied:
I tried different ways of building the graph and the easiest way I found was to start the graph from the right side = latest, filling up the weeks with no commits until the latest and from there go back to the earliest.

cloudhead commented on src/base/seeds/Seed.ts:
```diff
@@ -100,12 +100,14 @@ export class Seed {
-    return result.map((project: proj.ProjectInfo) => ({ ...project, id: project.urn }));
+    return Promise.all(
+      result.map((project: proj.ProjectInfo) => proj.Project.get(project.urn, null, null, this.api.host, config))
+    );
```
I think it would be best to separate these api queries, otherwise we have to wait for lots of requests to complete before the page can render.

sebastinez replied:
Ok reverted this to just map the urn to `id`

cloudhead commented on src/base/orgs/View/Projects.svelte:
```diff
@@ -59,25 +60,27 @@
-                  <AnchorActions
-                    {account} {config} anchor={pendingAnchor} safe={profile.org.safe}
-                    on:success={() => loadAnchors()} />
+          {#await project.getCommits(project.head) then points}
```
We shouldn't be waiting for this to load for the widget to render, ideally this is set on the widget once it is done loading in the background.

I also thought we wanted to only use the last 52 weeks of commits?

sebastinez replied:
> Another thing is that using this method forces us to fetch the project before getting its history, which shouldn't be necessary. Some options would be:

Yeah makes sense 👍 

> I also thought we wanted to only use the last 52 weeks of commits?

Yeah this escaped me, will implement it.

sebastinez replied:
Ok added the last 52 weeks of commits, and moved `getCommits` to be a class method like
`static async getCommits(urn: string, host: api.Host, parent: string, since?: string, until?: string): Promise<CommitsHistory>`

sebastinez commented:
Also found a edge case when a project didn't have a commit in the last year the svg path wasn't drawed because of a calculation error in the normalization formula

cloudhead commented on src/base/projects/Widget.svelte:
```diff
@@ -81,37 +106,62 @@
+    <div class="id">
+      <span class="urn desktop">{project.urn}</span>
+    </div>
+    {#if widgetHover}
```
Can CSS not be used for this? `article:hover .anchor { display: block }` `article:hover .activity { display: none }`?

cloudhead commented:
Made some tweaks to the visuals, can you apply this?

```diff
diff --git a/src/Diagram.svelte b/src/Diagram.svelte
index a80326a..72ece40 100644
--- a/src/Diagram.svelte
+++ b/src/Diagram.svelte
@@ -75,9 +75,7 @@
   <defs>
     <linearGradient id="gradient" x1="0" y1="1" x2="0" y2="0">
       <stop offset="0%" stop-color={fillColor} stop-opacity="0" />
-      <stop offset="10%" stop-color={fillColor} stop-opacity="0.1" />
-      <stop offset="25%" stop-color={fillColor} stop-opacity="0.2" />
-      <stop offset="50%" stop-color={fillColor} stop-opacity="0.3" />
+      <stop offset="100%" stop-color={fillColor} stop-opacity="0.4" />
     </linearGradient>
   </defs>
   <g>
diff --git a/src/base/projects/Widget.svelte b/src/base/projects/Widget.svelte
index da0c3d2..18008ae 100644
--- a/src/base/projects/Widget.svelte
+++ b/src/base/projects/Widget.svelte
@@ -43,6 +43,7 @@
   }
   .activity {
     width: 100%;
+    max-width: 14rem;
   }
   article:hover {
     border-color: var(--color-secondary);
@@ -154,9 +155,9 @@
         </span>
       </div>
     {:else}
-      <div class="activity">
+      <div class="activity desktop">
         <Diagram
-          strokeWidth={2}
+          strokeWidth={1.25}
           points={groupedCommits}
           viewBoxHeight={100}
           viewBoxWidth={600}

```

sebastinez commented:
> Made some tweaks to the visuals, can you apply this?
> 
> ```diff
> diff --git a/src/Diagram.svelte b/src/Diagram.svelte
> index a80326a..72ece40 100644
> --- a/src/Diagram.svelte
> +++ b/src/Diagram.svelte
> @@ -75,9 +75,7 @@
>    <defs>
>      <linearGradient id="gradient" x1="0" y1="1" x2="0" y2="0">
>        <stop offset="0%" stop-color={fillColor} stop-opacity="0" />
> -      <stop offset="10%" stop-color={fillColor} stop-opacity="0.1" />
> -      <stop offset="25%" stop-color={fillColor} stop-opacity="0.2" />
> -      <stop offset="50%" stop-color={fillColor} stop-opacity="0.3" />
> +      <stop offset="100%" stop-color={fillColor} stop-opacity="0.4" />
>      </linearGradient>
>    </defs>
>    <g>
> diff --git a/src/base/projects/Widget.svelte b/src/base/projects/Widget.svelte
> index da0c3d2..18008ae 100644
> --- a/src/base/projects/Widget.svelte
> +++ b/src/base/projects/Widget.svelte
> @@ -43,6 +43,7 @@
>    }
>    .activity {
>      width: 100%;
> +    max-width: 14rem;
>    }
>    article:hover {
>      border-color: var(--color-secondary);
> @@ -154,9 +155,9 @@
>          </span>
>        </div>
>      {:else}
> -      <div class="activity">
> +      <div class="activity desktop">
>          <Diagram
> -          strokeWidth={2}
> +          strokeWidth={1.25}
>            points={groupedCommits}
>            viewBoxHeight={100}
>            viewBoxWidth={600}
> ```

Ok addressed all the open issues, just one thing, here
```diff
diff --git a/src/base/projects/Widget.svelte b/src/base/projects/Widget.svelte
index da0c3d2..18008ae 100644
--- a/src/base/projects/Widget.svelte
+++ b/src/base/projects/Widget.svelte
@@ -154,9 +155,9 @@
         </span>
       </div>
     {:else}
-      <div class="activity">
+      <div class="activity desktop">
         <Diagram
           strokeWidth={1.25}
           points={groupedCommits}
           viewBoxHeight={100}
           viewBoxWidth={600}
```

The definition of `desktop` is
```css
.desktop {
  display: block !important;
}
```
if we add this here, we won't have the hover functionality and just the anchor showing up when hovering over the widget, is this your idea?

sebastinez commented:
> My intention was just to hide the activity graph on mobile, since it overlaps with everything else. Am I missing something?

Okay did some additional changes to your diff, in short:
- The Widget was resizing its height when hovering, I increased the min-height to 140px.
- There was an issue with using the same class `anchor` for the anchor actions and the `stateHash` so I added some differentiation.
- I extracted the linear gradient from the Diagram svg and put it inside `App.svelte` this on purpose so we can reuse the same gradient inside different components and it doesn't screw with the Widget layout.
- I added an `!important` tag to the `article:hover .activity` selector, since the `desktop` class would else overwrite the toggling visbility behaviour by hovering over the widget.

Let me know what you think, but I think we are close 😄

cloudhead commented on src/base/projects/LinearGradient.svelte:
```diff
@@ -0,0 +1,13 @@
+        <stop offset="100%" stop-color={fillColor} stop-opacity="0.4" />
+      </linearGradient>
+    </defs>
+  </svg>
```
Indentation :eyes:

cloudhead commented on src/base/orgs/View/Projects.svelte:
```diff
@@ -67,17 +75,19 @@
-                  <AnchorActions
-                    {account} {config} anchor={pendingAnchor} safe={profile.org.safe}
-                    on:success={() => loadAnchors()} />
+          {#await Project.getCommits(project.urn, seed.api, project.head, getTimestampOneYearAgo()) then points}
```
I think we discussed this, but I'm not too happy with waiting for the commit data to load before we are able to show the widget. Can this be moved so that the widget renders without waiting?

cloudhead commented on src/App.svelte:
```diff
@@ -16,6 +16,7 @@
   import Header from '@app/Header.svelte';
   import Loading from '@app/Loading.svelte';
   import Modal from '@app/Modal.svelte';
+  import LinearGradient from "@app/base/projects/LinearGradient.svelte";
```
If this is to be used as a global gradient, then maybe move it to `@app/LinearGradient.svelte`?
